apiserver/httpcontext.go

```diff
@@ -214,41 +215,34 @@ func (a tagKindAuthorizer) Authorize(authInfo httpcontext.AuthInfo) error {
 type controllerAuthorizer struct{}
 
 // Authorize is part of the httpcontext.Authorizer interface.
-func (controllerAuthorizer) Authorize(authInfo httpcontext.AuthInfo) error {
+func (controllerAuthorizer) Authorize(authInfo authentication.AuthInfo) error {
 	if authInfo.Controller {
 		return nil
 	}
 	return errors.Errorf("%s is not a controller", names.ReadableString(authInfo.Entity.Tag()))
 }
 
 type controllerAdminAuthorizer struct {
-	entityHasPermission func() entityHasPermissionFunc
-	controllerTag       names.Tag
+	controllerTag names.Tag
 }
 
 // Authorize is part of the httpcontext.Authorizer interface.
-func (a controllerAdminAuthorizer) Authorize(authInfo httpcontext.AuthInfo) error {
-	if authInfo.Token != nil {
-		access, err := permissionFromToken(authInfo.Token, a.controllerTag)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		if access != permission.SuperuserAccess {
-			return errors.Errorf("%s is not a controller admin", names.ReadableString(authInfo.Entity.Tag()))
-		}
-	}
-	if a.entityHasPermission == nil {
-		return errors.New("no permission checker configured")
-	}
+func (a controllerAdminAuthorizer) Authorize(authInfo authentication.AuthInfo) error {
 	userTag, ok := authInfo.Entity.Tag().(names.UserTag)
 	if !ok {
 		return errors.Errorf("%s is not a user", names.ReadableString(authInfo.Entity.Tag()))
 	}
-	isControllerAdmin, err := a.entityHasPermission()(userTag, permission.SuperuserAccess, a.controllerTag)
+
+	has, err := common.HasPermission(
+		common.UserAccessFunc(func(_ names.UserTag, subject names.Tag) (permission.Access, error) {
+			return authInfo.Permissions(subject)
```

Comment:
We should check the user tag arg to `common.UserAccessFunc` equals userTag above (for safety)